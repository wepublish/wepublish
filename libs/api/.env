DATABASE_URL=postgresql://postgres@localhost:5432/wepublish_test?schema=public