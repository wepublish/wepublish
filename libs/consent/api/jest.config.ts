/* eslint-disable */
export default {
  displayName: 'consent-api',
  preset: '../../../jest.preset.js',
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json'
    }
  },
  testEnvironment: 'node',
  transform: {
    '^.+\\.[tj]s$': 'ts-jest'
  },
  moduleFileExtensions: ['ts', 'js', 'html'],
  coverageDirectory: '../../../coverage/libs/consent/api',
  globalSetup: '<rootDir>/setup-database.js'
}
