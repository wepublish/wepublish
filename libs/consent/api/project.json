{
  "name": "consent-api",
  "$schema": "../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/consent/api/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/consent/api/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/consent/api/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  },
  "tags": []
}
