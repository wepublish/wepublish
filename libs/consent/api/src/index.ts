export * from './lib/consent.module'
