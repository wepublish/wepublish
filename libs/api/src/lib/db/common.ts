import {PageInfo} from '@wepublish/utils/api'

export interface ConnectionResult<T> {
  nodes: T[]
  pageInfo: PageInfo
  totalCount: number
}

export const DefaultSessionTTL = 1000 * 60 * 60 * 24 * 7 // 1w
export const DefaultBcryptHashCostFactor = 11

export enum DateFilterComparison {
  GreaterThan = 'gt',
  GreaterThanOrEqual = 'gte',
  Equal = 'eq',
  LowerThan = 'lt',
  LowerThanOrEqual = 'lte'
}
export interface DateFilter {
  date: Date | null
  comparison: DateFilterComparison
}
