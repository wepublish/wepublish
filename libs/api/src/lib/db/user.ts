import {
  PaymentProviderCustomer,
  User,
  UserAddress,
  UserOAuth2Account,
  MetadataProperty
} from '@prisma/client'
import bcrypt from 'bcrypt'
import {DefaultBcryptHashCostFactor} from './common'

export const hashPassword = async (
  password: string,
  bcryptHashCostFactor: number = DefaultBcryptHashCostFactor
) => bcrypt.hash(password, bcryptHashCostFactor)

export enum UserSort {
  CreatedAt = 'createdAt',
  ModifiedAt = 'modifiedAt',
  Name = 'name',
  FirstName = 'firstName'
}

export interface UserFilter {
  readonly name?: string
  readonly text?: string
  readonly userRole?: string[]
}

export type UserWithRelations = User & {
  address: UserAddress | null
  properties: MetadataProperty[]
  oauth2Accounts: UserOAuth2Account[]
  paymentProviderCustomers: PaymentProviderCustomer[]
}
