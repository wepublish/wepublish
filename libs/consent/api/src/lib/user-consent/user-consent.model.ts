import {ArgsType, Field, ObjectType} from '@nestjs/graphql'
import {User as UserType} from '@prisma/client'
import {User} from '@wepublish/user/api'
import {Consent} from '../consent/consent.model'

@ObjectType()
export class UserConsent {
  @Field()
  id!: string

  @Field()
  createdAt!: Date

  @Field()
  modifiedAt!: Date

  @Field(type => Consent)
  consent!: Consent

  @Field(type => User)
  user!: UserType

  @Field()
  value!: boolean
}

@ArgsType()
export class UserConsentInput {
  @Field()
  consentId!: string

  @Field()
  userId!: string

  @Field()
  value!: boolean
}

@ArgsType()
export class UserConsentFilter {
  @Field({nullable: true})
  name?: string

  @Field({nullable: true})
  slug?: string

  @Field({nullable: true})
  value?: boolean
}
