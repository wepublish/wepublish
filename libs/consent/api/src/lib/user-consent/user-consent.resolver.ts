import {Args, Mutation, Query, Resolver} from '@nestjs/graphql'
import {
  UserConsent,
  UserConsentInput,
  UpdateUserConsentInput,
  UserConsentFilter
} from './user-consent.model'
import {UserConsentService} from './user-consent.service'

import {UseGuards} from '@nestjs/common'
import {UserSession, AuthenticationGuard, CurrentUser} from '@wepublish/authentication/api'

@Resolver()
export class UserConsentResolver {
  constructor(private userConsents: UserConsentService) {}

  /*
  Queries
 */
  @Query(returns => [UserConsent], {
    name: 'userConsents',
    description: `
      Returns a list of userConsents. Possible to filter.
    `
  })
  userConsentList(@Args('filter', {nullable: true}) filter: UserConsentFilter) {
    return this.userConsents.userConsentList(filter)
  }

  @Query(returns => UserConsent, {
    name: 'userConsent',
    description: `
      Returns a single userConsent by id.
    `
  })
  userConsent(@Args('id') id: string) {
    return this.userConsents.userConsent(id)
  }

  /*
  Mutations
 */
  @Mutation(returns => UserConsent, {
    name: 'createUserConsent',
    description: `
      Creates a new userConsent based on input.
      Returns created userConsent.
    `
  })
  @UseGuards(AuthenticationGuard)
  createUserConsent(
    @CurrentUser() user: UserSession,
    @Args('userConsent') userConsent: UserConsentInput
  ) {
    // only allow creating for admin or affected user
    if (!user.user.roleIDs.includes('admin') && user.user.id !== userConsent.userId) {
      throw Error(`Unauthorized`)
    }
    return this.userConsents.createUserConsent(userConsent, user)
  }

  @Mutation(returns => UserConsent, {
    name: 'updateUserConsent',
    description: `
      Updates an existing userConsent based on input.
      Returns updated userConsent.
    `
  })
  @UseGuards(AuthenticationGuard)
  updateUserConsent(
    @CurrentUser() user: UserSession,
    @Args('id') id: string,
    @Args('userConsent') userConsent: UpdateUserConsentInput
  ) {
    return this.userConsents.updateUserConsent({id, userConsent, user})
  }

  @Mutation(returns => UserConsent, {
    name: 'deleteUserConsent',
    description: `
      Delete an existing userConsent by id.
      Returns deleted userConsent.
    `
  })
  @UseGuards(AuthenticationGuard)
  deleteUserConsent(@Args('id') id: string, @CurrentUser() user: UserSession) {
    return this.userConsents.deleteUserConsent(id, user)
  }
}
