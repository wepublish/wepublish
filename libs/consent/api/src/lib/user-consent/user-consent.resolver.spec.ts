import {ApolloDriver, ApolloDriverConfig} from '@nestjs/apollo'
import {INestApplication} from '@nestjs/common'
import {GraphQLModule} from '@nestjs/graphql'
import {Test, TestingModule} from '@nestjs/testing'
import {URLAdapter, URLAdapterModule} from '@wepublish/nest-modules'
import request from 'supertest'
import {UserConsentResolver} from './user-consent.resolver'
import {UserConsentService} from './user-consent.service'
import {UserDataloaderService} from '@wepublish/user/api'
import {createMock, PartialMocked} from '@wepublish/testing'

const userConsentQuery = `
  query userConsent($id: String!) {
    userConsent(id: $id) {
      id
      value
      createdAt
      modifiedAt
      consent {
        slug
        id
        name
      }
      user {
        __typename
        id
      }
    }
  }
`

const createUserConsentMutation = `
  mutation createUserConsent($consentId: String!, $userId: String!, $value: Boolean!) {
    createUserConsent(consentId: $consentId, userId: $userId, value: $value) {
      id
      value
    }
  }
`

const updateUserConsentMutation = `
  mutation updateUserConsent($id: String!, $value: Boolean!) {
    updateUserConsent(id: $id, value: $value) {
      id
      value
      createdAt
      modifiedAt
      consent {
        slug
        id
        name
      }
      user {
        __typename
        id
      }
    }
  }
`

const deleteUserConsentMutation = `
  mutation deleteUserConsent($id: String!) {
    deleteUserConsent(id: $id) {
      id
    }
  }
`

const mockUserConsent = {
  createdAt: new Date('2023-01-01'),
  id: 'userConsentId',
  modifiedAt: new Date('2023-01-01'),
  userId: 'userId',
  value: false,
  consent: {
    id: 'consentId',
    createdAt: new Date('2023-01-01'),
    modifiedAt: new Date('2023-01-01'),
    name: 'name',
    slug: 'slug',
    defaultValue: false
  }
}

const mockUser = {
  id: 'userId',
  name: 'name',
  firstName: 'firstName',
  birthday: null,
  email: 'email',
  active: true,
  flair: 'flair',
  userImageID: 'userImageId',
  roleIDs: []
}

const mockUserSession = {
  type: 'user',
  id: '448c86d8-9df1-4836-9ae9-aa2668ef9dcd',
  token: 'some-token',
  user: mockUser
}

describe('UserConsentResolver', () => {
  let app: INestApplication
  let userConsentService: PartialMocked<UserConsentService>
  let userDataloaderService: PartialMocked<UserDataloaderService>

  beforeEach(async () => {
    userConsentService = createMock(UserConsentService)
    userDataloaderService = createMock(UserDataloaderService)

    const module: TestingModule = await Test.createTestingModule({
      imports: [
        GraphQLModule.forRoot<ApolloDriverConfig>({
          driver: ApolloDriver,
          autoSchemaFile: true,
          path: '/',
          cache: 'bounded',
          context: {
            req: {
              user: mockUserSession
            }
          }
        }),
        URLAdapterModule.register(new URLAdapter(`https://example.com`))
      ],
      providers: [
        UserConsentResolver,
        {
          provide: UserConsentService,
          useValue: userConsentService
        },
        {
          provide: UserDataloaderService,
          useValue: userDataloaderService
        }
      ]
    }).compile()

    app = module.createNestApplication()
    await app.init()
  })

  afterAll(async () => {
    await app.close()
  })

  test('user consent query', async () => {
    const idToGet = 'id-to-get'

    userConsentService.userConsent?.mockResolvedValue(mockUserConsent)
    userDataloaderService.load?.mockResolvedValue(mockUser)

    await request(app.getHttpServer())
      .post('')
      .send({
        query: userConsentQuery,
        variables: {
          id: idToGet
        }
      })
      .expect(200)
      .expect(res => {
        expect(res.body).toMatchSnapshot()
      })
  })

  test('create user consent mutation', async () => {
    userConsentService.createUserConsent?.mockResolvedValue(mockUserConsent)

    await request(app.getHttpServer())
      .post('/')
      .send({
        query: createUserConsentMutation,
        variables: {
          consentId: 'consentId',
          userId: 'userId',
          value: true
        }
      })
      .expect(200)
      .expect(res => {
        expect(userConsentService.createUserConsent?.mock.calls).toMatchSnapshot()
        expect(res.body).toMatchSnapshot()
      })
  })

  test('update user consent mutation', async () => {
    userConsentService.updateUserConsent?.mockResolvedValue(mockUserConsent)
    userDataloaderService.load?.mockResolvedValue(mockUser)

    await request(app.getHttpServer())
      .post('/')
      .send({
        query: updateUserConsentMutation,
        variables: {
          id: 'userConsentId',
          value: false
        }
      })
      .expect(200)
      .expect(res => {
        expect(userConsentService.updateUserConsent?.mock.calls).toMatchSnapshot()
        expect(res.body).toMatchSnapshot()
      })
  })

  test('delete user consent mutation', async () => {
    userConsentService.deleteUserConsent?.mockResolvedValue(mockUserConsent)

    await request(app.getHttpServer())
      .post('/')
      .send({
        query: deleteUserConsentMutation,
        variables: {
          id: 'userConsentId'
        }
      })
      .expect(res => {
        expect(userConsentService.deleteUserConsent?.mock.calls).toMatchSnapshot()
        expect(res.body).toMatchSnapshot()
      })
      .expect(200)
  })
})
