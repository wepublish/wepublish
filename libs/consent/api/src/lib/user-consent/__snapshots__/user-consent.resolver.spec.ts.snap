// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserConsentResolver create user consent mutation 1`] = `
[
  [
    {
      "consentId": "consentId",
      "userId": "userId",
      "value": true,
    },
  ],
]
`;

exports[`UserConsentResolver create user consent mutation 2`] = `
{
  "data": {
    "createUserConsent": {
      "id": "userConsentId",
      "value": false,
    },
  },
}
`;

exports[`UserConsentResolver delete user consent mutation 1`] = `
[
  [
    "userConsentId",
    {
      "id": "448c86d8-9df1-4836-9ae9-aa2668ef9dcd",
      "token": "some-token",
      "type": "user",
      "user": {
        "active": true,
        "birthday": null,
        "email": "email",
        "firstName": "firstName",
        "flair": "flair",
        "id": "userId",
        "name": "name",
        "roleIDs": [],
        "userImageID": "userImageId",
      },
    },
  ],
]
`;

exports[`UserConsentResolver delete user consent mutation 2`] = `
{
  "data": {
    "deleteUserConsent": {
      "id": "userConsentId",
    },
  },
}
`;

exports[`UserConsentResolver update user consent mutation 1`] = `
[
  [
    "userConsentId",
    false,
    {
      "id": "448c86d8-9df1-4836-9ae9-aa2668ef9dcd",
      "token": "some-token",
      "type": "user",
      "user": {
        "active": true,
        "birthday": null,
        "email": "email",
        "firstName": "firstName",
        "flair": "flair",
        "id": "userId",
        "name": "name",
        "roleIDs": [],
        "userImageID": "userImageId",
      },
    },
  ],
]
`;

exports[`UserConsentResolver update user consent mutation 2`] = `
{
  "data": {
    "updateUserConsent": {
      "consent": {
        "id": "consentId",
        "name": "name",
        "slug": "slug",
      },
      "createdAt": "2023-01-01T00:00:00.000Z",
      "id": "userConsentId",
      "modifiedAt": "2023-01-01T00:00:00.000Z",
      "user": {
        "__typename": "User",
        "id": "userId",
      },
      "value": false,
    },
  },
}
`;

exports[`UserConsentResolver user consent query 1`] = `
{
  "data": {
    "userConsent": {
      "consent": {
        "id": "consentId",
        "name": "name",
        "slug": "slug",
      },
      "createdAt": "2023-01-01T00:00:00.000Z",
      "id": "userConsentId",
      "modifiedAt": "2023-01-01T00:00:00.000Z",
      "user": {
        "__typename": "User",
        "id": "userId",
      },
      "value": false,
    },
  },
}
`;
